### DISCUSSION TITLE
Prioritazation of backlog and new features

### DISCUSSION THREAD

**Discussion by fsheikh** [jgaeddert/liquid-dsp#263] [General]
Like in any other open source project it can be difficult for new-comers to gauge which feature/bug-fix/improvement really brings value to the project versus cosmetic changes. It would be nice to initiate a wish-list of things-to-do in priority order, with categories like:

- Completely new features, aligned to Liquid's original design/intentions which will enhance library's functionality
- Improvements like test-coverage of components, guidelines on how to write new tests, considerations for using tools/libraries like [fuzzing](https://llvm.org/docs/LibFuzzer.html)
- Examples, stuff that lives in sandbox to increase liquid's range, e.g. programs that may have nothing to do with digital communications but serve as applications in other domains. IMHO this is a good way to introduce liquid to users looking for a low-level signal processing library in domains like audio signal processing. A candidate example is [here](https://github.com/jgaeddert/liquid-dsp/pull/153). Disclaimer: Obviously its one of my patches :smile:

**Comment by jgaeddert** [comment]
Thanks for this, @fsheikh. From my perspective, the priorities in development are (in no particular order)

  1. Full [code coverage](https://app.codecov.io/gh/jgaeddert/liquid-dsp) (or greater than 90% within the next several months)
  2. Making the interface for objects more consistent (naming convention for objects, method names, variable ordering)
  3. Improving the usability on existing functions and methods (more/improved convenience methods, processing in blocks as opposed to having to write loops externally)
  4. Better documentation (mainly keeping the documentation up to date with the software)
  5. Binding the core C functionality up to [C++ and Python](https://github.com/jgaeddert/liquid-dsp/tree/bind-dev) for developers who want higher-level language options

I have a ton of local branches for new features that I've tinkered with, but I really want to get the existing structure solidified before adding more untested features.

**Reply by pfeatherstone** [reply]
I'm wondering if it wouldn't be better to refactor to C++ and provide C bindings, as well as python bindings. (python bindings to C++ is super easy with pybind11). Now hear me out:
- With C++, you could template the input type and parameter types. So you would get rid of all macro nastiness which I hope everyone agrees is bad and a pre-processor hack
- Therefore the library would be largely header-only, nice! No build step!
- No more compiler warning explosion related to ```liquid_complex_float``` and ```std::complex<float>``` not being convertible by default if using clang.
- You wouldn't need fancy features, pretty much just RAII and maybe some inheritance if you wanted to consolidate APIs like ```push()```, ```execute()```, ```execute_block()``` into a base class, though not necessary. So really C++98 would be sufficient. Surely most people who use liquid have a C++98 compiler if they have a C compiler. I would struggle to believe otherwise.
- A lot of memory related bugs would disappear. Yes, you wouldn't need a single (!) call to ```malloc()```, ```free()```, ```new``` or ```delete```. Buffers could be allocated with ```std::vector``` and if for any reason you need pointers to objects, then use a smart pointer and either ```make_shared()``` or ```make_unique```. Though in that case C++11 would be required. But i really don't think this is necessary. RAII should be sufficient. And if C++11 is available, you could use move semantics to minimise copies. But copy semantics are fine too.
- API misuse would be substantially reduced because of the stronger type system in C++, proper use of objects and well defined object lifetime and pointer casting would be either a compiler error or at least a warning. 
- With RAII you substantially reduce your code. At least every single function like ```liquid_object_create()``` and ```liquid_object_destroy()``` would disappear
- Using STL algorithms like ```std::transform```, ```std::copy``` etc, I foresee a lot of vector operations like dot products would be auto-vectorized. Yes, have you ever looked at what gcc vectorizes? Standard C-like for loops are almost never vectorized whereas STL algorithms are!
- Proper object alignment. Indeed, have you ever noticed that every object in liquid is allocated using ```malloc()``` (there is no other way of course) and that doesn't give a damn about object alignment. You would probably be ok because it likely returns memory on a 8-byte boundary and that's usually fine but i don't think there is anything in the standard that guarantees that. Whereas an object allocated on the stack using RAII is always properly aligned. Just saying... Wouldn't be surprised if someone somewhere has experienced undefined behaviour on some dodgy platform because of this.
- A C API around the C++ API would be fairly easy. A lot of libraries do that to inter-op with higher level languages like Rust, Go and so-on. 
- A python API would be trivial with pybind11.

What do people think? I don't want to insight a language war between old school C programmers and C++ programmers. It just seems that a lot of bugs and library misuses would go away... At the end of the day:
- C++ is a much safer language
- If written correctly, i.e. use the standard library and its algorithms, its performance surpasses C
- It's still a comparatively low level language so nobody should get too upset
- Much easier to build cleaner and safer APIs

I'm sure there are a billion more reasons. Then again, I'm sure some C developers would have reasons to keep liquid in pure C too.

**Reply by pfeatherstone** [reply]
I'm not convinced this would be too much work by the way. It seems like a good opportunity to do a bit of a re-write given everything else that's going on now.

**Reply by JayKickliter** [reply]
I understand all the arguments for C++ above, and don't disagree with them in theory, exceot for `I'm not convinced this would be too much work by the way`. It'll be lot of work for not a whole lot of gain. From experience (sorry for the appeal to authority), when binding the lib to other languages, the C wrappers with be a bottle neck compared to directly binding to the original pure C lib. Finally, the lib already works very well, is very clean, and jgaeddert has stated his strong preference for C over C++.

**Reply by pfeatherstone** [reply]
With respect to the amount of work, all the hard stuff, i.e. algorithms, is done and tested. The work is mainly refactoring. So yeah, don't think that's too much work. 

I don't think calling a C binding to C++ implementation is going to be a bottleneck. I imagine your C binding will just be returning pointers to the underlying C++ objects and the C-style functional API will just be calling the public member functions of the C++ objects. So it's not like there is even an indirection. E.g.:

```
class SomeObject
{
public:
    void execute(...);
};

typedef some_object_t = SomeObject*;

extern "C" {
some_object_t create() {return new SomeObject();}
void destroy(some_object_t obj) {delete obj;}
void execute(some_object_t obj) {obj->execute();}
}
```
I can't see how the above is going to be a bottleneck.

However, I agree the library works well as-is and is very clean. But as with every C library, it's incredibly easy to use incorrectly and the compiler does diddly squat to prevent that.

**Reply by pfeatherstone** [reply]
I've lost track of the number of issues related to buffer overflows or stack overflows in liquid over the years. Using an STL container would give you a runtime exception when compiled in debug and using ```_GLIBCXX_DEBUG```

**Reply by pfeatherstone** [reply]
But I don't know, i can understand this would seem like going backwards coz it's a complete re-write. But in my head its worth it in order to avoid accruing any more technical debt. But I'm a big fan of C++ and library design, so I'm biased.

**Reply by pfeatherstone** [reply]
Last comment about this I swear. A good library should be literally impossible to use incorrectly. A misuse should either be a compile time error or a very expressive and immediate runtime error. The next best thing is something that is very hard to use incorrectly. I think a C++ rewrite would allow for this.

**Reply by fsheikh** [reply]
The advantages of converting C++ cannot be denied but like everything, nothing is free. I can think of paying the price in following terms.

- Size of the library, templates will inevitably increase the size of generated code. I am not sure how many consumers rely on fitting `liquid` on a few MBs of ROM.
- There are many C++ signal processing/communication libraries already out-there. At a cursory look [1](https://github.com/mohabouje/eDSP) and [2](https://github.com/stefanseefeld/openvsip) look to have similar objective. Point is `cppliquid` may end up replicating an existing solution with very little gain in functionality. 
- I also agree that some kinds of errors are avoidable by simply switching to C++, but they are not eliminated. For C programmers moving to C++ it is common to still write loop based indexing as apposed to iterators which does not take stack overflows completely out of picture. Also errors may be much harder to debug, specially in the case of `shared_pointers`, with cycles of shared ownership (without breaking via `weak_pointer`) resulting in memory leaks, again much harder to debug. 

Having said that, I will be very interested to convert one component e.g. one of the `modems` to C++ and see some performance/size/memory consumption results for this component.

**Reply by jgaeddert** [reply]
I realize this wasn't really ever advertised, but has anyone tested the [bind-dev](https://github.com/jgaeddert/liquid-dsp/tree/bind-dev) branch? Binding to [C++](https://github.com/jgaeddert/liquid-dsp/tree/bind-dev#c-bindings) and [Python3](https://github.com/jgaeddert/liquid-dsp/tree/bind-dev#python-bindings) is fairly straightforward, I think. I'm using `pybind11` as @pfeatherstone suggests which makes instantiating the framing classes pretty clean:

```python
#!/usr/bin/env python3
import liquid as dsp, numpy as np, matplotlib.pyplot as plt 

# generate a frame
fg      = dsp.fg64()
header  = np.random.randint(256, size=fg.header_len,  dtype=np.uint8)
payload = np.random.randint(256, size=fg.payload_len, dtype=np.uint8)
frame   = fg.execute(header, payload)

def callback(context,header,payload,stats):
    print('frame detected!')
    context.update(stats)

# create frame synchronizer and receive frame
context = {'syms':np.zeros((0,))}
fs = dsp.fs64(callback,context)
fs.execute(frame)
```

**Reply by brian-armstrong** [reply]
Just to blow this up a bit, I actually think there could be huge value in rewriting everything into Rust. The build system feels so much nicer than trying to get C++ to build on every platform, and Liquid is quite nicely heavily tested so that the tests could be translated and used to verify everything. Rust also has answers for a lot of the issues that come up, even if it does have a fairly steep learning curve.

I started to rewrite libcorrect into Rust but I got a bit sidetracked by trying to make the asm perform well. Otherwise it felt like a very good language to use. There's also some strength in capitalizing on the current momentum behind Rust imo, and Liquid would be a fantastic first-choice DSP library for that language.

**Reply by jgaeddert** [reply]
Also, the C++ binding (really, wrapping) is header-only, so there's nothing additional to compile once the core C library is built.

```c++
// frame gen/sync 64 C++ example          
#include <stdio.h>
#include "fg64.hh"
#include "fs64.hh"
        
namespace dsp = liquid;
        
int main()
{       
    // generate header and payload bytes
    uint8_t header[8], payload[64];
    for (auto i=0; i< 8; i++) header[i]  = rand() & 0xff;
    for (auto i=0; i<64; i++) payload[i] = rand() & 0xff;

    // create frame generator and write frame to buffer
    dsp::fg64 fg;
    std::cout << fg << std::endl;
    unsigned int frame_len = fg.get_frame_length();
    std::complex<float> frame_buf[frame_len];
    fg.execute(header, payload, frame_buf);

    // create frame synchronizer and receive frame
    dsp::fs64 fs(NULL,NULL);
    std::cout << fs << std::endl;
    fs.execute(frame_buf, frame_len);
    printf("number of packets detected: %u\n", fs.get_framedatastats().num_frames_detected);
    return 0; 
}
```

I realize this isn't an ideal solution, but as several of you have pointed out, C++ _is_ a lot easier to use in a lot of ways. The memory protection, etc. comes at the cost of speed, though, which is why I avoided a lot of the containers in the standard template library to begin with. Maybe for the best, maybe not.

**Reply by jgaeddert** [reply]
@brian-armstrong Rust is a fascinating suggestion

**Reply by pfeatherstone** [reply]
@jgaeddert You've noticed a slowdown in using STL containers? That's very surprising. I've noticed the opposite! You have to write in a C++ way though.

**Reply by pfeatherstone** [reply]
This might be a loosing battle I think. C developers will claim C++ is slower and C++ developers will say C is slower. Probably because one is doing the other badly... :)

**Reply by pfeatherstone** [reply]
In any case, this has been interesting. It looks like a pure C implementation is probs best. And people can always wrap the liquid objects if they want to. I'm sure everyone on this thread has done this countless times. There's no need for liquid to be opinionated on the matter. It can just carry on providing a solid C API. And in fact, providing a c++ API might not be the best use of resources as everyone will have a different opinion on how to wrap it and may end up writing their own.

**Reply by pfeatherstone** [reply]
Maybe it's best to focus on @jgaeddert's first 4 points in his initial comment.

**Reply by jgaeddert** [reply]
It's true. I'm sure C vs C++ can be faster depending upon how they're used, particularly if they're used improperly.

**Comment by pfeatherstone** [comment]
I would also add to the list:
- use cmake build system and provide a liquid.cmake file so ```find_package()``` works
- use Catch2 for unit testing

**Reply by JayKickliter** [reply]
I reluctantly prefer CMake too. I hate it, but less than I hate most other build systems that have reasonable adoption. In the meantime, we liquid can still install a `liquid.cmake`.

**Reply by pfeatherstone** [reply]
I used to hate CMake but since using it properly it's quite nice, at least way nicer than anything else, as you said. It's literally two files: ```CMakeLists.txt``` and ```liquid.cmake```. It also means integrating in other projects is trivial. And you can use things like ```FetchContent()``` so building dependencies doesn't require polluting your system wide installations.

**Reply by JayKickliter** [reply]
Off-topic: FetchContent seemed so great, but we had to stop using it at my company. Lots of systems out there still have older versions of CMake. Theoretically, we could upgrade them, but the folks in charge of those systems don't know anything about CMake and would be stuck for while trying to figure out how to fix the failing build.

**Reply by pfeatherstone** [reply]
Fair enough. But I can understand not being able to update your compiler for whatever reason, but I can't understand not being able to update cmake. You can download the binary from the website and runs everywhere.

**Reply by jgaeddert** [reply]
I will say, I hate autoconf.

**Comment by pfeatherstone** [comment]
I've had a look at the C++ bindings. I have a couple issues:
- the wrappers don't define copy constructors or move constructors. The fact that the copy constructors aren't defined and/or  explicitly marked as deleted is a huge bug since objects which are copied will be pointing to the same underlying C object. So you will have double frees when copies are destructed. 

I suggest using something like this:

```
struct liquid_deleter
{
    void operator()(modem obj)          { if (obj) modem_destroy(obj); }
    void operator()(fskmod obj)         { if (obj) fskmod_destroy(obj); }
    void operator()(firfilt_cccf obj)   { if (obj) firfilt_cccf_destroy(obj); }
    void operator()(msequence obj)      { if (obj) msequence_destroy(obj); }
    ...
};

template<typename LiquidObject>
using liquid_ptr = std::unique_ptr<typename std::remove_pointer<LiquidObject>::type, liquid_deleter>;
```

Then something like:
```
liquid_ptr<modem>
```
is totally fine and well defined and can be used instead of ```modem``` inside the C++ wrapper. The wrapper does NOT need to define any of the following methods: destructor, copy constructor, copy assign, move constructor, move assign, swap.

- Some objects have loads of constructors, a lot of which have default parameters. I think users will have many bugs when calling the wrong constructors due to implicit casting of primitive types. I think if you're going to have many constructors, use [named-constructors](https://isocpp.org/wiki/faq/ctors#named-ctor-idiom)

**Reply by jgaeddert** [reply]
Agreed. This is something I've struggled with for a while and where C++ has some obvious benefits over C. Copying the underlying data structures is difficult as they often retain state that isn't trivial to duplicate. C++ can do this recursively, of course, but it's much more of a manual process in C. Which isn't to say that it's impossible; just more difficult.

**Reply by pfeatherstone** [reply]
In which case for now, implicitly disable copying by using ```liquid_ptr<>``` as suggested above. At least, destruction, swap and move semantics will be correct for free.

**Reply by pfeatherstone** [reply]
I think this is where a pure C++ implementation is advantageous. All this stuff is handled by the type system. And if you want to use it in C, just write a C binding. Seems backwards I know, but tons of libraries do this (x265, zeromq, etc)

**Reply by jgaeddert** [reply]
This is one of the reasons I've started this process here; I wanted to convert several different classes of functionality to get a feel for how easy or difficult it is to convert. Filter design requires many different types of arguments (having a named constructor makes sense here as you pointed out). The arbitrary rate resamplers are tricky as they require varying output lengths. This can be handled by std::vector but might not be the most elegant solution.

**Reply by pfeatherstone** [reply]
I've actually written an arbitrary rate resampler from scratch in pure C++. The code size is about 20% of the liquid code and it is between 2 and 10 times faster (depending on the rates), and that's without using SIMD or threading. Just plain old C++ and the STL. (compiling with -O2).

**Reply by pfeatherstone** [reply]
Oh, and without FFTW

**Reply by pfeatherstone** [reply]
I think there is an argument for using exceptions in the C++ binding for error handling rather than using ```liquid_error_config```. Indeed ```liquid_error_config``` is typically used for bad API usage, which is by definition exceptional behavior. For non-exceptional behavior, then return codes are the right choice.

**Reply by jgaeddert** [reply]
The way I see it, there's only so much flexibility and convenience that C provides (outside of portability, speed of compiling, etc.). Wrapping in C++ classes and binding to Python is a path forward, but is slow and very much a manual process. Outside of re-writing everything in C++ (as others have suggested--for better or for worse), are there idealized ways to support C/C++/Python without manually keeping everything up to date? C++ and Python are fairly easy with pybind11, but I don't see a clear path for supporting both C and C++.

**Reply by pfeatherstone** [reply]
Yeah I would just choose one and stick to it. Though I prefer C++ to C, C is a better common denominator for supporting potential bindings to other languages (Rust, Go, Swift, Emscripten, etc). Though I wouldn't make providing bindings a priority. Other repositories can do that if they want. To be honest, getting the docs up to date and fixing bugs are higher priority in my view.

**Reply by pfeatherstone** [reply]
I find the liquid docs very educational and very useful!

**Reply by brian-armstrong** [reply]
I will yet again add a vote for a rewrite in Rust, if anything. C to Rust bindings are relatively straightforward, and Rust is a modern language with a very portable toolchain. And Liquid would be able to capitalize on the Rust momentum quite well. The memory safety aint bad either!

**Reply by jgaeddert** [reply]
@brian-armstrong I've never used rust, but you've got me intrigued!
